servicetalk-http-api/src/main/java/io/servicetalk/http/api/HeaderUtils.java

```diff
@@ -667,12 +667,18 @@ static boolean hasContentType(final HttpHeaders headers,
             return false;
         }
 
-        if (UTF_8.equals(expectedCharset) &&
-                ((contentEqualsIgnoreCase(expectedContentType, TEXT_PLAIN) &&
-                        contentEqualsIgnoreCase(contentTypeHeader, TEXT_PLAIN_UTF_8)) ||
-                (contentEqualsIgnoreCase(expectedContentType, APPLICATION_X_WWW_FORM_URLENCODED) &&
-                        contentEqualsIgnoreCase(contentTypeHeader, APPLICATION_X_WWW_FORM_URLENCODED_UTF_8)))) {
-            return true;
+        if (UTF_8.equals(expectedCharset)) {
+            if (!hasCharset(contentTypeHeader)) {
+                return true;
+            }
+            if (contentEqualsIgnoreCase(expectedContentType, TEXT_PLAIN) &&
+                    contentEqualsIgnoreCase(contentTypeHeader, TEXT_PLAIN_UTF_8)) {
+                return true;
+            }
+            if (contentEqualsIgnoreCase(expectedContentType, APPLICATION_X_WWW_FORM_URLENCODED) &&
+                    contentEqualsIgnoreCase(contentTypeHeader, APPLICATION_X_WWW_FORM_URLENCODED_UTF_8)) {
```

Comment:
This is preexisting, but I think we don't need the full `contentEqualsIgnoreCase` here. My suggestion will be to add a length check as the first check inside `if (UTF_8.equals(expectedCharset))`. Because we already verified that we match the type and subtype in line 666, that will help to avoid unnecessary `hasCharset` and first `contentEqualsIgnoreCase`. 

The second suggestion will be to use `regionMatches` to find `; charset=UTF-8`. Because this is a fastlane shortcut, we should not worry about various OWS combinations, they will be handled after this `if`.

The final code for fastlane will be:
```java
if (UTF_8.equals(expectedCharset)) {
    if (contentTypeHeader.length() == expectedContentType.length()) {
        return true;
    }
    if (regionMatches(contentTypeHeader, true, expectedContentType.length(), "; charset=UTF-8", 0, 15)) {
        return true;
    }
    if (!hasCharset(contentTypeHeader)) {
        return true;
    }
}
```